### PR TITLE
Add generic templates for DataResponse and JSONResponse

### PR DESCRIPTION
**Pull request by provokateurin**
## Summary

Add templates to indicate generic type. Not required, but very useful for OpenAPI.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by come-nc** [comment]
This adds a new constraints that one cannot use setData with another type than what was used in the constructor. No idea if this is a problem or not.

File: lib/public/AppFramework/Http/JSONResponse.php [hunk 1]
```diff
@@ -32,18 +32,19 @@
 /**
  * A renderer for JSON calls
  * @since 6.0.0
+ * @template T
  */
 class JSONResponse extends Response {
 	/**
 	 * response data
-	 * @var array|object
+	 * @var T
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/JSONResponse.php:40]: this weakens the type. before it was an array or object. now it can be anything

**Comment by provokateurin** [comment]
Waiting on a solution for https://github.com/nextcloud/server/issues/36513 before continuing since this change might not be needed after all.

**Comment by provokateurin** [comment]
I just realized this isn't possible, since psalm will complain about missing template types in every place that isn't updated.

**Pull request closed without merging**